From 8482caa4ec57477d57be5cb26a2380dda644746b Mon Sep 17 00:00:00 2001
From: Pradeep Kumar Chitrapu <quic_pradeepc@quicinc.com>
Date: Fri, 18 Nov 2022 14:34:19 -0800
Subject: [PATCH] ath12k: drop packets with invalid peer id in dp_tx path

Sometime DP might receive a packet with already deleted peer id.
Added a check to avoid sending such packets.

Signed-off-by: Pradeep Kumar Chitrapu <quic_pradeepc@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/dp_tx.c | 14 ++++++++++++++
 1 file changed, 14 insertions(+)

diff --git a/drivers/net/wireless/ath/ath12k/dp_tx.c b/drivers/net/wireless/ath/ath12k/dp_tx.c
index d7ceaf3133cf..5937088003cc 100644
--- a/drivers/net/wireless/ath/ath12k/dp_tx.c
+++ b/drivers/net/wireless/ath/ath12k/dp_tx.c
@@ -245,6 +245,7 @@ int ath12k_dp_tx(struct ath12k *ar, struct ath12k_vif *arvif,
 	struct ath12k_base *ab = ar->ab;
 	struct ath12k_hw_group *ag = ab->ag;
 	struct ath12k_dp *dp = &ab->dp;
+	struct ath12k_peer *peer;
 	struct hal_tx_info ti = {0};
 	struct ath12k_tx_desc_info *tx_desc = NULL;
 	struct ieee80211_tx_info *info = IEEE80211_SKB_CB(skb);
@@ -257,6 +258,7 @@ int ath12k_dp_tx(struct ath12k *ar, struct ath12k_vif *arvif,
 	void *hal_tcl_desc;
 	u8 hal_ring_id;
 	int ret;
+	u16 peer_id;
 	bool msdu_ext_desc = false;
 
 	if (unlikely(test_bit(ATH12K_FLAG_CRASH_FLUSH, &ar->ab->dev_flags)))
@@ -285,6 +287,18 @@ int ath12k_dp_tx(struct ath12k *ar, struct ath12k_vif *arvif,
 	    arsta->use_4addr_set) {
 		ti.meta_data_flags = arsta->tcl_metadata;
 		ti.flags0 |= FIELD_PREP(HAL_TCL_DATA_CMD_INFO2_TO_FW, 1);
+		spin_lock_bh(&ab->base_lock);
+		peer_id = FIELD_GET(HTT_TCL_META_DATA_PEER_ID, ti.meta_data_flags);
+		peer = ath12k_peer_find_by_id(ab, peer_id);
+		if (!peer || !peer->sta) {
+			ath12k_dbg(ab, ATH12K_DBG_DP_TX,
+				   "Dropped packet with non existent peer id %u\n", peer_id);
+			ret = -EINVAL;
+			spin_unlock_bh(&ab->base_lock);
+			goto fail_unmap_dma;
+		}
+		spin_unlock_bh(&ab->base_lock);
+
 	} else {
 		ti.meta_data_flags = arvif->tcl_metadata;
 	}
-- 
2.17.1

